Открываем сводную таблицу с помощью pandas (10 аннотаторов, 7 текстов)

In [8]:
import pandas as pd

df = pd.read_csv('int_annot_ggl.tsv', sep = '\t', index_col = 0)
df

text1      text2     text3     text4       text5  \
Annotator1   Нейронауки   Культура   История   История      Физика   
Annotator2   Нейронауки   Культура  Биология   История      Физика   
Annotator3   Нейронауки    История   История   История      Физика   
Annotator4   Нейронауки   Культура  Биология   История      Физика   
Annotator5   Нейронауки   Культура  Биология   История  Математика   
Annotator6   Нейронауки    История   История  Культура      Физика   
Annotator7   Нейронауки  Философия  Биология   История  Математика   
Annotator8   Нейронауки    История  Биология   История      Физика   
Annotator9   Нейронауки   Культура   История   История      Физика   
Annotator10  Нейронауки   Культура  Биология   История      Физика   

                  text6      text7  
Annotator1     Биология   Культура  
Annotator2     Биология   Культура  
Annotator3     Биология   Культура  
Annotator4   Психология   Культура  
Annotator5     Биология   Культура  
Annotator6     Биология   Культура  
Annotator7     Биология  Философия  
Annotator8   Нейронауки   Культура  
Annotator9     Биология   Культура  
Annotator10    Биология   Культура

Готовим данные для обработки: т.к. мы будем использовать модуль nltk.metrics.agreement, на вход нам потребуется список таплов (аннотатор, текст, лейбл):

In [9]:
annot_data = []

for text in df.columns:
    for coder in df.T.columns:
        annot_data.append((coder, text, df.T[coder][text]))
        

In [10]:
from nltk import agreement

ratingtask = agreement.AnnotationTask(data=annot_data)
print("kappa " +str(ratingtask.kappa()))
print("fleiss " + str(ratingtask.multi_kappa()))
print("alpha " +str(ratingtask.alpha()))
print("scotts " + str(ratingtask.pi()))

kappa 0.6111234479472432
fleiss 0.6053067993366503
alpha 0.6088044022011005
scotts 0.6031349007837253


Получаем:
- каппа Коэна (для двух аннотаторов, сравнивает их попарно) - 0.6111234479472429 
- каппа Фляйса (мульти-каппа Коэна, для нескольких аннотаторов) - 0.6053067993366503
- альфа коэффициент Крипендорфа - 0.6088044022011005
- пи Скотта - 0.6031349007837253

Вышеперечисленные статистические методы оценки согласованности между несколькими аннотаторами, классифицирующими n текстов по m категориям, выдают очень близкие значения - усредненно 0,6. Это хороший показатель, близкий к высокому (высокий показатель > 0,75)